py/mkrules.mk: Add dependency of .mpy files upon mpy-cross.

dpgeorge opened the pull request:
This ensures that mpy-cross is automatically built (and is up-to-date) for ports that use frozen bytecode.  It also makes sure that .mpy files are re-built if mpy-cross is changed.

I used a simple *.[ch] pattern match for getting the dependencies, which should be good enough.

@dhylands what do you think of it?

dpgeorge committed:
py/mkrules.mk: Add dependency of .mpy files upon mpy-cross.

This ensures that mpy-cross is automatically built (and is up-to-date) for
ports that use frozen bytecode.  It also makes sure that .mpy files are
re-built if mpy-cross is changed.

dhylands commented on py/mkrules.mk:
```diff
@@ -102,15 +102,19 @@ $(BUILD)/frozen.c: $(wildcard $(FROZEN_DIR)/*) $(HEADER_BUILD) $(FROZEN_EXTRA_DE
 
 ifneq ($(FROZEN_MPY_DIR),)
+# to build the MicroPython cross compiler
+$(TOP)/mpy-cross/mpy-cross: $(TOP)/py/*.[ch] $(TOP)/mpy-cross/*.[ch] $(TOP)/windows/fmode.c
```
Personally, I think I'd be inclined to just drop the dependenies and make $(TOP)/mpy-cross/mpy-cross be declared .PHONY so that it always runs make. Make should run pretty fast (less than a second on my machine) if it's up-to-date.

so something like:
```
.PHONY: $(TOP)/mpy-cross/mpy-cross
$(TOP)/mpy-cross/mpy-cross:
    $(Q)$(MAKE) -C $@
```
Then you don't run into any weird things where you forgot some dependency.

dhylands replied:
Well that didn't quite work the way I thought it would. Let me muddle this...

dhylands replied:
@dpgeorge So I think your original solution is the best one that works (my suggestion has some serious flaws). If I think of something else, I'll let you know.

dpgeorge commented:
Thanks for the feedback @dhylands .  I did try to do a similar thing to what you suggest, but it involved editting all the individual port's Makefile's to make their "all:" target depend on the phony "mpy-cross" target.  I prefer the solution that I posted here in this PR.

But I found a small problem: with ports like unix that have targets like "make coverage" that re-invoke make with certain flags/vars set, these flags/vars seem to get passed to the subsequent invocation of $(MAKE) that is used to compile mpy-cross.  In the case of unix's "make coverage" that leads to, for example, the setting `-DMP_CONFIGFILE="<mpconfigport_coverage.h>"` propagate through to mpy-cross, which gives an error.  How to get around this?

dhylands commented:
You can put an if statement around the code looking for particular targets or variables. See this for an example:
https://github.com/dhylands/projects/blob/cdbf7f705e8efa2e1a00607a2298d427c2aa6130/rules/mkRules.mk#L130-L134
that uses MAKECMDGOALS and filter to not include dependencies when doing a clean. (micropython does the dependencies slightly differently and didn't need this).

The variable MAKECMDGOALS is the goals (like deploy) passed on the command line to make.

You could also do a hybrid approach where you set some special variable (in the unix Makefile) which does the mpy-cross psuedo-target approach.

dpgeorge commented:
@dhylands sorry but I don't understand how your suggestion fixes the problem I mentioned.

I looked deeper into the issue and it's almost impossible to fix.  The issue is that any variables passed on the command line to recursive make invocations are set in the outer shell and therefore propagate to further recursive make calls, in the case at hand to the compilation of mpy-cross.  For example:
```
$ make coverage
--> calls make recursively with -DMP_CONFIGFILE="<mpconfigport_coverage.h>"
--> this makes a new shell environment with env variable MP_CONFIGFILE set to the above value
--> makefile calls make again to build mpy-cross 
--> MP_CONFIGFILE variable is still set and causes build of mpy-cross to fail
```
The only way around this (without drastically changing the unix/Makefile) is to "override undefine", in mpy-cross's Makefile, all the variables that have leaked from the coverage build into mpy-cross.  This seems like a very dirty hack that will lead to lots of other problems down the track.

dpgeorge committed:
mpy-cross/Makefile: Override undefine any inherited variables.

This is a temporary hack to allow othe Makefiles to build mpy-cross
recursively.

dpgeorge commented:
I pushed a commit to show how "override undefine" could work.

dhylands commented:
So are you trying to not build mpy-cross when building the coverage build?
Or are you trying to build a regular (non-coverage) mpy-cross when building the coverage build?

dpgeorge commented:
> Or are you trying to build a regular (non-coverage) mpy-cross when building the coverage build?

Yes.  mpy-cross is an independent program and doesn't care about coverage or x86 or xtensa etc. You can build one mpy-cross for all targets.  All I want to do is make sure mpy-cross is up-to-date before then compiling .py to .mpy for a given port.  Eg if a file in py/ changes then mpy-cross should be rebuilt.  `make -C mpy-cross` will do it but I want that to be automated.

pfalcon commented:
@dpgeorge , So first of all, the reason why we didn't do it before is because we anticipated the issues like that (I can't believe there was another reason). So, no surprise they happen. Here, just as in parallel discussion, we're hitting make's limitations.

Anyway, if the problem that make settings from one component link into completely unrelated component, then the obvious solution would be to stop the leakage. make's settings leak thru the environment, so the solution would be avoid running "submake" and instead to running an independent make. There would be multiple potential ways to achieve that, starting from looking for a suitable option in GNU make, and to robustly dumb solution of resetting the environment when running make for mpy-cross. Did you try that?

pfalcon commented:
> looking for a suitable option in GNU make

And that's not wishful thinking, I faced a similar problem in esp-open-sdk, where the whole thing is effectively a makefile script, and it calls ct-ng to build, which also uses makefile and unwisely picks up things from the driver makefile script. Instead of dumping the magic variable name, let me direct to look into it yourself, because I have no idea if it matches your scenario or not in the first place.

dpgeorge commented:
> so the solution would be avoid running "submake" and instead to running an independent make.

Yes I tried many ways of doing this, eg just doing `make` instead of `$(MAKE)`, right up to running a separate shell like `echo make | /bin/sh`.  The same problem persists, that env variables are always inherited to sub shells and hence sub make's.

pfalcon commented:
@dpgeorge : Tried `env -i`? Tried make `MAKELEVEL=0` like esp-open-sdk does?

dhylands commented:
I'd probably avoid using MAKELEVEL as it will change if somebody else builds upy make from some other make.

We could just have the top-level unix make set a MPY_CROSS_BUILT variable which then prevents any upy makes from trying to rebuild (i.e. omits the dependencies) in a submake.

dpgeorge commented:
> Tried env -i? Tried make MAKELEVEL=0 like esp-open-sdk does?

MAKELEVEL=0 doesn't work, the env variables are still set and used.

`env -i` does clear the relevant env variables but does too good a job and introduces a new error: gcc can no longer find the cc1 executable: `gcc: error trying to exec 'cc1': execvp: No such file or directory
`

pfalcon commented:
> MAKELEVEL=0 doesn't work, the env variables are still set and used.

Yes, they should be set, but should be not used. Why does it work for esp-open-sdk then? How exactly do  you apply it?

dpgeorge commented:
> How exactly do you apply it?

I just tried passing MAKELEVEL=0 in the recipe that builds mpy-cross, ie `$(Q)make -C $(TOP)/mpy-cross MAKELEVEL=0`.  AFAIX, to make use of MAKELEVEL you need to actually change the mpy-cross/Makefile to test the value of this variable and act differently depending on its value.  Eg you still need to use `override undefine` to clear all the incoming variables.

dpgeorge commented:
@pfalcon if there are no other ideas then I'd like to merge this as a temporary solution, because otherwise there are going to be lots of issues upgrading from 1.8.7 to 1.9, since mpy-cross needs to be rebuilt.

pfalcon commented:
@dpgeorge : Sounds good.
